packages/repository/src/repositories/relation.factory.ts

```diff
@@ -0,0 +1,164 @@
+}
+/**
+ * Enforces a constraint on a repository based on a relationship contract
+ * between models. Returns a relational repository that exposes applicable CRUD
+ * method APIs for the related target repository. For example, if a Customer model is
+ * related to an Order model via a HasMany relation, then, the relational
+ * repository returned by this method would be constrained by a Customer model
+ * instance's id(s).
+ *
+ * @param constraint The constraint to apply to the target repository. For
+ * example, {id: '5'}.
+ * @param relationMetadata The relation metadata used to used to describe the
+ * relationship and determine how to apply the constraint.
+ * @param targetRepository The repository which represents the target model of a
+ * relation attached to a datasource.
+ *
+ */
+export function constrainedRepositoryFactory<T extends Entity, ID>(
+  constraint: AnyObject,
+  relationMetadata: HasManyDefinition,
+  targetRepository: EntityCrudRepository<T, ID>,
+): HasManyEntityCrudRepository<T, ID> {
+  switch (relationMetadata.type) {
+    case RelationType.hasMany:
+      const fkConstraint: AnyObject = {};
+      fkConstraint[relationMetadata.keyTo] =
+        constraint[relationMetadata.keyFrom];
+
+      return new DefaultHasManyEntityCrudRepository(
+        targetRepository,
+        fkConstraint,
+      );
+  }
+}
+
+/**
+ * A utility function which takes a filter and enforces constraint(s)
+ * on it
+ * @param originalFilter the filter to apply the constrain(s) to
+ * @param constraint the constraint which is to be applied on the filter
+ * @returns Filter the modified filter with the constraint, otherwise
+ * the original filter
+ */
+export function constrainFilter(
+  originalFilter: Filter | undefined,
+  constraint: AnyObject,
+): Filter {
+  let constrainedFilter: Filter = {};
+  const constrainedWhere = new WhereBuilder();
+  for (const c in constraint) {
+    constrainedWhere.eq(c, constraint[c]);
+  }
+  if (originalFilter) {
+    constrainedFilter = cloneDeep(originalFilter);
+    if (originalFilter.where) {
+      constrainedFilter.where = constrainedWhere.and(
+        originalFilter.where,
+      ).where;
+    }
+  } else if (originalFilter === undefined) {
+    constrainedFilter.where = constrainedWhere.where;
+  }
+  return constrainedFilter;
+}
+
+/**
+ * A utility function which takes a filter and enforces constraint(s)
+ * on it
+ * @param originalFilter the filter to apply the constrain(s) to
+ * @param constraint the constraint which is to be applied on the filter
+ * @returns Filter the modified filter with the constraint, otherwise
+ * the original filter
+ */
+export function constrainWhere(
+  originalWhere: Where | undefined,
+  constraint: AnyObject,
+): Where {
+  const constrainedWhere = new WhereBuilder();
+  for (const c in constraint) {
+    constrainedWhere.eq(c, constraint[c]);
+  }
+  if (originalWhere) {
+    constrainedWhere.where = constrainedWhere.and(originalWhere).where;
```

Comment:
Can we improve `WhereBuilder` and `FilterBuilder` to provide the feature of parsing and updating an existing filter/where definition please? This will allow us to simplify these two methods to focus on what we need to achieve, instead of dealing with low-level details.

E.g.

```ts
// one option
const builder = WhereBuilder.fromQuery(originalWhere);
builder.assign(constraint);

// or perhaps
const builder = new WhereBuilder();
builder.mergeWithWhere(originalWhere);
builder.mergeWithKeyValueMap(constraint);
```

Reply:
+1 to move the helper methods to corresponding builders. For example:

```ts
const builder = WhereBuilder.from(originalWhere);
builder.impose(constraint); // or build.constrainWith(constraint)
```

Reply:
Looks like I overlooked passing the `originalWhere` in the constructor of `WhereBuilder`. Also, there is a function which does the `constrainWith` or `mergeWithKeyValueMap` (https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/query.ts#L163-L175). It's marked `private` right now, but it does form an `and` clause for conflicting keys. Maybe we can refactor it?